onyo/lib/onyo.py

```diff
@@ -1240,3 +1241,128 @@ def unset(self, paths: Iterable[Union[Path, str]], values: str,
             self.restore()
 
         return diff
+
+    #
+    # GET
+    #
+    # TODO: This should be a staticmethod in utils
+    def _walk(self, path, depth) -> Generator:
+        """Recursively yield filepaths up to a given depth"""
+        depth -= 1
+        with os.scandir(path) as p:
+            for entry in p:
+                # ignore hidden files and folders
+                # TODO: Better to use _is_protected_path or ignore hidden?
+                #   `if self._is_protected_path(entry.path):`
+                if entry.name.startswith('.'):
+                    continue
+
+                # yield discovered assets
+                if not entry.is_dir():
+                    yield entry.path
+
+                # explore dirs if depth allows
+                if entry.is_dir() and depth >= 0:
+                    yield from self._walk(entry.path, depth)
+
+    @staticmethod
+    def _filter_by_name(assets: List[str], filters: dict) -> Generator:
+        """Filter assets by filename to prevent redundant asset loading"""
+        keys = ['type', 'make', 'model', 'serial']
+        for asset in assets:
+            split = set(zip(keys, re.split('[_.]', os.path.basename(asset))))
```

Comment:
I think this could be a useful private method in its own right. Thoughts?

Also, please use `pathlib` rather than `os.path`.

Reply:
`pathlib` is the default when using `self.assets` instead of the now deprecated `self._walk` anyway, so `os` is completely out.

I'm also in favor of adding a method to split a filename into its type, make, model, and serial components. Why private, though?

Reply:
Giving this further thought, the core of the functionality comes down to `re.split('[_.]', asset)` (where `asset` is `Path.name` passed to the function). This is trivial, and makes no real sense to add as a function.

Instead I would suggest adding this to a future implementation of the `Asset` dataclass, instantiated using a `Path` object. The `__post_init__` could then automatically split `Path.name` into `Asset.typ`, `Asset.make`, `Asset.model`, and `Asset.serial` properties. These can then be modified if/when we make the asset file naming scheme customizable.